mailbox/event/event-rabbitmq/src/test/java/org/apache/james/mailbox/events/RabbitMQEventBusConcurrentTest.java

```diff
@@ -0,0 +1,285 @@
+import org.testcontainers.shaded.com.google.common.collect.ImmutableSet;
+
+import com.jayway.awaitility.core.ConditionFactory;
+
+class RabbitMQEventBusConcurrentTest {
+
+    private static class RabbitMQEventBusDeclarationExtension implements BeforeEachCallback, AfterEachCallback {
+
+        private EventSerializer eventSerializer;
+        private RabbitMQConnectionFactory connectionFactory;
+        private RoutingKeyConverter routingKeyConverter;
+
+        @Override
+        public void beforeEach(ExtensionContext context) throws Exception {
+            rabbitMQExtension.beforeEach(context);
+
+            connectionFactory = rabbitMQExtension.getConnectionFactory();
+
+            TestId.Factory mailboxIdFactory = new TestId.Factory();
+            routingKeyConverter = RoutingKeyConverter.forFactories(new MailboxIdRegistrationKey.Factory(mailboxIdFactory));
+            eventSerializer = new EventSerializer(mailboxIdFactory, new TestMessageId.Factory());
+        }
+
+        @Override
+        public void afterEach(ExtensionContext context) throws Exception {
+            rabbitMQExtension.afterEach(context);
+            ALL_LISTENERS.forEach(listener -> listener.clear());
+        }
+
+        private RabbitMQEventBus newEventBus() {
+            return new RabbitMQEventBus(connectionFactory, eventSerializer, routingKeyConverter);
+        }
+    }
+
+    private static final Duration FIVE_SECONDS = Duration.ofSeconds(5);
+    private static final ConditionFactory AWAIT_CONDITION = await().timeout(new com.jayway.awaitility.Duration(5, TimeUnit.SECONDS));
+    private static final ImmutableSet<RegistrationKey> NO_KEYS = ImmutableSet.of();
+    private static final EventBusContract.MailboxListenerCountingSuccessfulExecution COUNTING_LISTENER_1 =
+        new EventBusContract.MailboxListenerCountingSuccessfulExecution();
+    private static final EventBusContract.MailboxListenerCountingSuccessfulExecution COUNTING_LISTENER_2 =
+        new EventBusContract.MailboxListenerCountingSuccessfulExecution();
+    private static final EventBusContract.MailboxListenerCountingSuccessfulExecution COUNTING_LISTENER_3 =
+        new EventBusContract.MailboxListenerCountingSuccessfulExecution();
+    private static final List<EventBusContract.MailboxListenerCountingSuccessfulExecution> ALL_LISTENERS =
+        ImmutableList.of(COUNTING_LISTENER_1, COUNTING_LISTENER_2, COUNTING_LISTENER_3);
+
+    private static final int THREAD_COUNT = 10;
+    private static final int OPERATION_COUNT = 30;
+    private static final int TOTAL_DISPATCH_OPERATIONS = THREAD_COUNT * OPERATION_COUNT;
+
+    @RegisterExtension
+    static RabbitMQExtension rabbitMQExtension = new RabbitMQExtension();
+    @RegisterExtension
+    static RabbitMQEventBusDeclarationExtension testExtension = new RabbitMQEventBusDeclarationExtension();
+
+// TODO tests isolation problem makes assertions become wrong. Fix it then enable this trick
+//    @BeforeAll
```

Comment:
Commented code is never a good idea...

Reply:
This PR is in WIP mode, comment keeps me to not forget this issue, latter, comment will be removed.